File: apps/backend/serving/triage/slack.py
"""Small Slack Web API client used by the relay."""

from __future__ import annotations

import httpx


class SlackDeliveryError(RuntimeError):
    """Raised when Slack rejects or cannot receive a message."""


class SlackClient:
    """Post top-level messages and thread replies with a bot token."""

    def __init__(self, token: str, channel_id: str, *, timeout_seconds: float = 10.0) -> None:
        self._token = token
        self._channel_id = channel_id
        self._timeout_seconds = timeout_seconds

    async def post(self, text: str, *, thread_ts: str | None = None) -> str:
        """Post a message and return its Slack timestamp."""
        payload: dict[str, str] = {"channel": self._channel_id, "text": text[:40_000]}
        if thread_ts:
            payload["thread_ts"] = thread_ts
        try:
            async with httpx.AsyncClient(timeout=self._timeout_seconds) as client:
                response = await client.post(
                    "https://slack.com/api/chat.postMessage",
                    headers={"Authorization": f"Bearer {self._token}"},
                    json=payload,
                )
            response.raise_for_status()
            body = response.json()
        except (httpx.HTTPError, ValueError) as exc:
            raise SlackDeliveryError("Slack chat.postMessage request failed") from exc
        if not isinstance(body, dict):
            raise SlackDeliveryError("Slack chat.postMessage returned an invalid response")
        if not body.get("ok") or not body.get("ts"):
            error = str(body.get("error") or "unknown_error")
            raise SlackDeliveryError(f"Slack chat.postMessage rejected the request: {error}")
        return str(body["ts"])

File: services/alert-relay-worker/src/slack.ts
import type { SlackMessage } from "./types";

export class SlackDeliveryError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "SlackDeliveryError";
  }
}

export interface SlackWriter {
  postParent(message: SlackMessage, clientMessageId: string): Promise<string>;
  updateParent(threadTs: string, message: SlackMessage): Promise<void>;
  postReply(threadTs: string, message: SlackMessage, clientMessageId: string): Promise<void>;
}

interface SlackResponse {
  ok?: boolean;
  ts?: string;
  error?: string;
}

export class SlackApiClient implements SlackWriter {
  constructor(
    private readonly token: string,
    private readonly channelId: string,
    private readonly fetcher: typeof fetch = fetch,
  ) {}

  private async call(method: "chat.postMessage" | "chat.update", payload: object): Promise<SlackResponse> {
    let response: Response;
    try {
      response = await this.fetcher(`https://slack.com/api/${method}`, {
        method: "POST",
        headers: {
          Authorization: `Bearer ${this.token}`,
          "Content-Type": "application/json; charset=utf-8",
        },
        body: JSON.stringify(payload),
        signal: AbortSignal.timeout(10_000),
      });
    } catch {
      throw new SlackDeliveryError(`Slack ${method} request failed`);
    }
    if (!response.ok) {
      throw new SlackDeliveryError(`Slack ${method} returned HTTP ${response.status}`);
    }
    let body: SlackResponse;
    try {
      body = (await response.json()) as SlackResponse;
    } catch {
      throw new SlackDeliveryError(`Slack ${method} returned invalid JSON`);
    }
    if (!body.ok) {
      throw new SlackDeliveryError(`Slack ${method} rejected the request: ${body.error ?? "unknown"}`);
    }
    return body;
  }

  async postParent(message: SlackMessage, clientMessageId: string): Promise<string> {
    const response = await this.call("chat.postMessage", {
      channel: this.channelId,
      text: message.text,
      blocks: message.blocks,
      client_msg_id: clientMessageId,
      unfurl_links: false,
      unfurl_media: false,
    });
    if (!response.ts) throw new SlackDeliveryError("Slack chat.postMessage omitted ts");
    return response.ts;
  }

  async updateParent(threadTs: string, message: SlackMessage): Promise<void> {
    await this.call("chat.update", {
      channel: this.channelId,
      ts: threadTs,
      text: message.text,
      blocks: message.blocks,
    });
  }

  async postReply(
    threadTs: string,
    message: SlackMessage,
    clientMessageId: string,
  ): Promise<void> {
    await this.call("chat.postMessage", {
      channel: this.channelId,
      thread_ts: threadTs,
      text: message.text,
      blocks: message.blocks,
      client_msg_id: clientMessageId,
      reply_broadcast: false,
      unfurl_links: false,
      unfurl_media: false,
    });
  }
}
